add dlink dir-1960-a1 and dir-2660-a1 support

bendavid opened the pull request:
This continues from https://github.com/openwrt/openwrt/pull/3210 with the device trees for the dir-1960-a1 and dir-2660-a1 based on https://github.com/Lucky1openwrt/openwrt/commit/0886f9ccb629cde4add485950144f9a0e5f7567f

The only change for the existing devices (dir-867,877,882) aside the reorganization of the common device trees, is the setting of the default mac addresses for the wireless interfaces in the common device tree.

bendavid committed:
ramips: add support for D-Link DIR-2660 A1

This patch adds support for D-Link DIR-2660 A1. Given
the similarity with the DIR-1760/1960 A1, this patch also introduces a common DTS
which can be shared with these devices, with support to be added in future commits.

Specifications:
* Board: AP-MTKH7-0002
* SoC: MediaTek MT7621AT
* RAM: 256 MB (DDR3)
* Flash: 128 MB (NAND)
* WiFi: MediaTek MT7615N (x2)
* Switch: 1 WAN, 4 LAN (Gigabit)
* Ports: 1 USB 2.0, 1 USB 3.0
* Buttons: Reset, WPS
* LEDs: Power (white/orange), Internet (white/orange), WiFi 2.4G (white),
        WiFi 5G (white), USB 3.0 (white), USB 2.0 (white)

Notes:
* WiFi 2.4G and WiFi 5G LEDs are wired directly to the wireless chips

Installation:
* D-Link Recovery GUI: power down the router, press and hold the reset
  button, then re-plug it. Keep the reset button pressed until the power
  LED starts flashing orange, manually assign a static IP address under
  the 192.168.0.xxx subnet (e.g. 192.168.0.2) and go to http://192.168.0.1

* Some modern browsers may have problems flashing via the Recovery GUI,
  if that occurs consider uploading the firmware through cURL:

    curl -v -i -F "firmware=@file.bin" 192.168.0.1

Signed-off-by: Josh Bendavid <joshbendavid@gmail.com>

bendavid commented:
I only have access to the dir-2660-a1 myself, so if e.g. @mbc07 @Lucky1openwrt @Sixtiz could help with testing this on the other devices, that would be great.

adrianschmutzler commented:
Your submission has formal issues, please visit https://openwrt.org/submitting-patches and have a look at earlier similar contributions.

If you have only one device tested, please split the commit into two separate commits (within this PR).

adrianschmutzler commented:
Looking at the device-tree, it might actually be easier to create a DTSI for your devices only, and not touch target/linux/ramips/dts/mt7621_dlink_dir-8xx-a1.dtsi at all. This will also allow you to keep the keys in the DTSI ...

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-1960-a1.dts:
```diff
@@ -0,0 +1,68 @@
+		led-boot = &led_power_orange;
+		led-failsafe = &led_power_white;
+		led-running = &led_power_white;
+		led-upgrade = &led_net_orange;
```
I don't think using the net LED is wise here, as a blinking net LED can be interpreted differently.

mbc07 replied:
This actually follows the behavior of the stock firmware, which also blinks the orange net LED while updating the firmware.

adrianschmutzler replied:
Will this actually work, i.e. will the DT setup really overwrite the user-space setup as LAN led when upgrading?

Despite, this would only make sense if you limit the net LED to the link state and remove blinking on rx/tx.

mbc07 replied:
It works fine because the orange net LED is unused on OpenWrt apart from indicating firmware upgrade. Link on, RX and TX activity are mapped to the green net LED (which apart from being white on the DIR-1760/1960/2660 is the same GPIO).

adrianschmutzler replied:
Sorry, you are right and I was talking nonsense (partly mixed this up with another PR).

Might only be a problem when a user sets up something to that LED manually.

I'd personally still not use a LED designated for ethernet here when there are actually two "power" LEDs available. But that's the author's design choice now.

mbc07 replied:
I'd personally leave as it is. It matches the stock firmware behavior (both for booting and also for upgrading), works out of the box with the default OpenWrt settings and it's exactly how it's currently configured on the DIR-867/878/882 DTS. And parting from the premise of "the user might manually setup the LED to do something different", the same is valid for any of the available LEDs, including the power LEDs and the WiFi LEDs...

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-1960-a1.dts:
```diff
@@ -0,0 +1,68 @@
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
```
This has been moved to mt7621.dtsi and can be dropped.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-1960-a1.dts:
```diff
@@ -0,0 +1,68 @@
+		};
+
+		usb3_white {
+			label = "dir-1960-a1:white:usb3.0";
```
Is there an USB 2.0 LED anywhere? Otherwise just name this "usb".

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-1960-a1.dts:
```diff
@@ -0,0 +1,68 @@
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
```
drop empty line.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-2660-a1.dts:
```diff
@@ -0,0 +1,75 @@
+		};
+
+		usb3_white {
+			label = "dir-2660-a1:white:usb3.0";
```
I'd just use "usb2" and "usb3" here, without the minor revision.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-xx60-a1.dtsi:
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 shouldn't be here: https://openwrt.org/docs/guide-developer/device-support-policies#dts-v1

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -232,6 +232,36 @@ define Device/dlink_dir-8xx-a1
+  UBINIZE_OPTS := -E 5
+  DEVICE_VENDOR := D-Link
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware \
+  	kmod-usb3 kmod-usb-ledtrig-usbport wpad-basic
```
Indent one tab only.

mbc07 reviewed (commented):
I don't have access to the DIR-867 anymore (it was borrowed and I already returned it), but I will test this PR on my DIR-882 soon...

adrianschmutzler commented:
> I don't have access to the DIR-867 anymore (it was borrowed and I already returned it), but I will test this PR on my DIR-882 soon...

Note that I actually requested to not touch these devices here.

mbc07 commented:
So, I tested this on DIR-882 (just in case) and nothing was broken. However, I do agree that leaving DIR-867/878/882 alone, with their own shared DTS and having a separate one for the DIR-1760/1960/2660 trio seems better. The way it is currently is a bit confusing as you have to jump over 3 different files to read the entire DTS for one of these devices.

Given all of them use the same board, I would probably add the DIR-1760 here too. From the device tree perspective, it is identical to the DIR-1960, the only difference being the number of spatial streams of the WiFi modules but that's handled automatically by the mt76 driver (just like on the DIR-867). Jumping from the DIR-1760/1960 to the DIR-2660 the difference is an additional USB 2.0 port and one more USB LED...

Lucky1openwrt commented:
DIR-878 & DIR-1960 Basic test work fine :)
sysupgrade & Factory from recovery page &, simple 5G WiFi

bendavid committed:
ramips: add support for D-Link DIR-1960 A1

This patch adds support for D-Link DIR-1960 A1.

Specifications:
* Board: AP-MTKH7-0002
* SoC: MediaTek MT7621AT
* RAM: 256 MB (DDR3)
* Flash: 128 MB (NAND)
* WiFi: MediaTek MT7615N (x2)
* Switch: 1 WAN, 4 LAN (Gigabit)
* Ports: 1 USB 3.0
* Buttons: Reset, WPS
* LEDs: Power (white/orange), Internet (white/orange), WiFi 2.4G (white),
        WiFi 5G (white), USB 3.0 (white)

Notes:
* WiFi 2.4G and WiFi 5G LEDs are wired directly to the wireless chips

Installation:
* D-Link Recovery GUI: power down the router, press and hold the reset
  button, then re-plug it. Keep the reset button pressed until the power
  LED starts flashing orange, manually assign a static IP address under
  the 192.168.0.xxx subnet (e.g. 192.168.0.2) and go to http://192.168.0.1

* Some modern browsers may have problems flashing via the Recovery GUI,
  if that occurs consider uploading the firmware through cURL:

    curl -v -i -F "firmware=@file.bin" 192.168.0.1

Signed-off-by: Josh Bendavid <joshbendavid@gmail.com>

bendavid commented:
PR updated with suggested changes.  (In particular this is now decoupled from the dir-867,877,882, though probably setting the mac addresses for the wireless interface also makes sense there.)

I can easily add the DIR-1760 as well if someone would be able to test it.

mbc07 commented:
> I can easily add the DIR-1760 as well if someone would be able to test it.

AFAICT DIR-1760 is in the same situation of DIR-867: very hard to find nowadays since both were discontinued last year and they weren't released globally, only in some countries. However, given the DTS for it would be identical to the DIR-1960 (apart from changing the references of "DIR-1960" to "DIR-1760") and considering that all other D-Link devices that uses the AP-MTKH7-0002 board already were tested and confirmed working, I see no reason to **not** include the DIR-1760 on this PR, even if untested...

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -232,6 +232,29 @@ define Device/dlink_dir-8xx-a1
+  DEVICE_VENDOR := D-Link
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware \
+	kmod-usb3 kmod-usb-ledtrig-usbport wpad-basic
+  KERNEL := $$(KERNEL) | uimage-padhdr 96
```
Does this really need to be applied to the KERNEL here?

bendavid replied:
As opposed to where?

adrianschmutzler replied:
Well, for dir-8xx it is applied to everything except the kernel ...

bendavid replied:
If I understood correctly it's equivalent given that uimage-padhdr is applying to the beginning of whatever is there so far.

Lucky1openwrt replied:
you could change the dir-8xx if you like but both work 
  KERNEL := $$(KERNEL) | ubootpad96
  IMAGES += factory.bin
  IMAGE/sysupgrade.bin := append-kernel | append-rootfs |\
	 append-metadata | check-size
  IMAGE/factory.bin := append-kernel | append-rootfs | check-size

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-xx60-a1.dtsi:
```diff
@@ -0,0 +1,167 @@
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		ieee80211-freq-limit = <5000000 6000000>;
+		mtd-mac-address = <&factory 0xe000>;
+		mtd-mac-address-increment = <2>;
```
Are there no addresses provided by caldata, i.e. in 0x4 and 0x8004?

Lucky1openwrt replied:
Not on the DIR-1960 just random numbers that don't match the one's used int the OEM firmware "00223344AA00" & "000C12006028" in mine 
should be "60 63 4C 66 E4 EC" & "60 63 4C 66 E4 ED"

adrianschmutzler replied:
So, you have addresses in 0xe000 and 0xe006 but not in 0x4 and 0x8004? That's interesting, quite uncommon.

In that case, I'd be interested in an overview of how addresses _should_ be on that device:
https://openwrt.org/docs/guide-developer/device-support-policies#documentation

That could be added to the commit message for that device then.

Same should be done for the 2660 as well.

Lucky1openwrt replied:
DIR-1960
lan          factory 0xe000   *:EB (label)
wan        factory 0xe006   *:EE
2.4 GHz  factory 0xe000+1 *:EC
5.0 GHz  factory 0xe000+2 *:ED
seems like they didn't replace the dummy entry's in the calibration data
I was expecting the same as the DIR-878 but it was not

bendavid replied:
Yes I'll have a look.

Lucky1openwrt replied:
just need to look at you MTD2 file in a hex editor at stated locations
I'm happy to look if you can get me your MTD2 file

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-1960-a1.dts:
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
This file needs to be removed for 2660-a1 ...

adrianschmutzler commented:
checkpatch.pl told me something about Windows/DOS line endings in the DTS(I) files.

Despite, in the commit message: "introduces a common DTS" -> "introduces a common DTSI"

adrianschmutzler commented:
Since MAC address assignment was available for 1960 already, I decided to cut that out and merge it already.
Note that I moved the WiFi MAC assignment to the DTS file, so technically the "default" from DTSI would still be the data from calibration data.

adrianschmutzler commented:
Ping. Please update the PR for 2660 support and check the MAC addresses on that device.

bendavid commented:
Yes, not forgotten, will get to it in the next few days.

adrianschmutzler commented:
I've done the rebase myself and pushed it to my staging tree:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

I just need the feedback on the MAC address assignment now.

newbs commented:
Hi, what needs to happen for the DIR-2660 to appear in the snapshots - https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/

I have a DIR-2660 in the post - is there anything that I can do to facilitate support for the DIR-2660?

Can the DIR-1960 images be used on the DIR-2660 hardware?

adschm commented:
> Hi, what needs to happen for the DIR-2660 to appear in the snapshots - https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/
> 
> I have a DIR-2660 in the post - is there anything that I can do to facilitate support for the DIR-2660?

Tell me the MAC addresses of the device with vendor firmware (i.e. the last byte) for lan/wan/2g/5g.

bendavid commented:
Unfortunately was waiting for me to find some time to flash back the factory firmware and check in detail the default mac address behaviour.

adschm commented:
@bendavid Can you provide a hexdump -C of the factory partition (from openwrt)?
@newbs Should still have vendor firmware, so I just need a table with the MAC addresses on factory from him.

bendavid commented:
Sure, here it is:

[apswitch2.mtd2.bin.zip](https://github.com/openwrt/openwrt/files/5164105/apswitch2.mtd2.bin.zip)

adschm commented:
> Sure, here it is:
> 
> [apswitch2.mtd2.bin.zip](https://github.com/openwrt/openwrt/files/5164105/apswitch2.mtd2.bin.zip)

Okay, so the stored addresses are like dir-1960:

0xe000 *:18
0xe006 *:1b

nothing useful in 0x4 and 0x8004

So now, we just need the OEM assignment; and as a bonus, it would be nice to get the label MAC address from one or both of you.

Lucky1openwrt commented:
> Can the DIR-1960 images be used on the DIR-2660 hardware?

I have tested the DIR-2660 on the DIR-1960 & it just runs & looks the same only model Number difference
if you need a decrypted firmware to flash via the recover console
http://luckys.onmypc.net/openwrt/DIR-2660/DIR2660A1_FW104B03.bin 
from this you can flash the normal dlink encrypted files

newbs commented:
> So now, we just need the OEM assignment; and as a bonus, it would be nice to get the label MAC address from one or both of you.

for the OEM assignments do you need the hex file of the MTD2 section (if so any tips on retrieving this)?

for the label MAC this the start of the MAC address on the router label C4:E9:0A:xx:xx:A7

on the DLINK DIR-2660 firmware web interface the WAN (Internet) MAC is listed as above but ends C4:E9:0A:xx:xx:AA

apologies if any of the above seems noobie - I deal more with Cortex M series parts than MPU & linux etc. but hopefully i am giving you the correct information.

Lucky1openwrt commented:
lan factory 0xe000 *:A7 (label)
wan factory 0xe006 *:AA
2.4 GHz factory 0xe000+1 *:A8
5.0 GHz factory 0xe000+2 *:A9
I'm going to guess this is it but can you confirm your OEM firmware matches these @newbs

newbs commented:
> lan factory 0xe000 *:A7 (label)
> wan factory 0xe006 *:AA
> 2.4 GHz factory 0xe000+1 *:A8
> 5.0 GHz factory 0xe000+2 *:A9
> I'm going to guess this is it but can you confirm your OEM firmware matches these @newbs
> the idea is to make openwrt MAC addresses match the ones in the OEM firmware for the same interface

I have an issue with the firmware recovery page.  

The luckys.onmypc link isn't working for me so I downloaded the 1.04b firmware from d-link and used that to downgrade to. i then grabbed the latest 1960 snapshot to flash to the router.  this boots (and i can log in over SSH) but from the D'Link recovery page i cannot revert to the D'link firmware.

If i can get back to stock firmware I was then going to use inSSIDer to check what the wifi MACs were but I am now stuck...

Lucky1openwrt commented:
@newbs
http://luckys.onmypc.net/openwrt/DIR-2660/DIR2660A1_FW104B03.bin
maybe try a different browser it works here on firefox\win 10\

bendavid commented:
I never got this working from a browser in linux and ended up using a windows vm+usb NIC with usb passthrough.

You can also use curl as explained in the commit message.

Lucky1openwrt commented:
form his comment he can't download the firmware form my link 
if he can upload openwrt he should be able to do the dlink firmware the same way

newbs commented:
@Lucky1openwrt the link was an issue with my anti virus...
@bendavid it appears that not all windows 10 version of chrome, firefox or opera are created equal, tried with a second laptop on chrome and the upgrade immediately proceeded as it normally does (ie as I am used to on the DIR878 devices).

@adschm 2.4Ghz = *A8, 5Ghz  = *A9.

Lucky1openwrt commented:
@newbs
lan factory 0xe000 *:A7 (label)
wan factory 0xe006 *:AA
2.4 GHz factory 0xe000+1 *:A8
5.0 GHz factory 0xe000+2 *:A9
are you confirming all 4 MAC addresses match the above & are the same in both d-link & openwrt firmware ?

newbs commented:
my previous observations were all on the D'link firmware.
Running the 1960 openwrt snapshot and ebaling radio 0 and radio1 ifconfig gives:
eth0 = *A7
wlan0 = *A8
wlan1 = *A9
wan = *AA

these match with the D'link firmware.

Lucky1openwrt commented:
@newbs 
thank you
I be leave this was the last piece of information @adschm was after
👍

adschm commented:
Thanks, dir-2660 will be merged soon.

mbc07 commented:
For future reference, I discovered a new router from D-Link (DIR-2640 A1), hardware is absolutely identical to the DIR-2660 A1 apart from the white LEDs being blue instead (orange LEDs unchanged) and the Wi-Fi toggle push button being reintroduced after its absence from the DIR-1760/1960/2660 trio.

Stock firmware layout between DIR-2660 and DIR-2640 is also identical, the only noticeable difference is the introduction of Google Assistant and Alexa integration on the DIR-2640, while the McAfee protection from DIR-2660 was axed on DIR-2640...

basrieter commented:
Is there any chance the flashing of image can be fixed? Newer bootloaders require some additional headers in kernel image. I reported this here: https://bugs.openwrt.org/index.php?do=details&task_id=3641

It seems like a small fix, but I don't even know where to start in OpenWRT?
